feat(perf-issues): Add Large Render Blocking Asset docs (#6338)
getsentry/sentry-docs: src/docs/product/issues/issue-details/performance-issues/large-render-blocking-asset/index.mdx

---
title: "Large Render Blocking Asset"
sidebar_order: 30
redirect_from:
  - /product/issues/performance-issues/large-render-blocking-asset/
description: "Learn more about large render blocking assets and how to diagnose and fix them."
---

> <Include name="early-adopter-note.mdx" />

Reviewer: ```suggestion
<Include name="early-adopter-note.mdx" />
```
Could you add this feature to the EA list here: https://docs.sentry.io/product/accounts/early-adopter/

Author: Good catch, done!

> A _Large Render Blocking Asset_ is a performance problem that happens when a large asset causes a delay in displaying the page content. For example, if a page includes a stylesheet, the browser will pause rendering of the page until it downloads and processes the entire stylesheet. If the stylesheet is large, the user may see a blank or unstyled screen for a long time. This problem occurs when applications do not split up asset files, load non-essential assets synchronously, or load assets too early.
> 

Reviewer: ```suggestion
A _Large Render Blocking Asset_ is a performance problem that happens when a large asset causes a delay in displaying the page content. For example, if a page includes a stylesheet, and the browser pauses page rendering until the the entire stylesheet downloads and processes. If that stylesheet is large, the user may see a blank or unstyled screen for a long time. This problem occurs when applications don't split up asset files, load non-essential assets synchronously, or load assets too early.

```

## Detection Criteria

> The detector for this performance issue looks for specific asset (resource) spans with the following criteria:
> 
> - The span operation has to be one of either: `resource.link` or `resource.script`
> - The span's `resource.render_blocking_status` data property must have a value of `blocking` (added by the `@sentry/tracing` SDK as of version `7.38.0`) or be missing.
> - The span's `Encoded Body Size` data property has to be over `1MB` in size
> - The [FCP (First Contentful Paint)](/product/performance/web-vitals/#first-contentful-paint-fcp) of the transaction must be at least `2` seconds (with an upper bound of `10` seconds to eliminate outliers).
> - The span's duration must be at least `33%` of the FCP

Reviewer: ```suggestion
The detector for this performance issue looks for specific asset (resource) spans with the following criteria:

- The span operation has to be either a: `resource.link` or `resource.script`.
- The span's `resource.render_blocking_status` data property should either have a value of `blocking` (added by the `@sentry/tracing` SDK, version `7.38.0`), or be missing.
- The span's `Encoded Body Size` data property has to be over `1MB`.
- The [First Contentful Paint (FCP)](/product/performance/web-vitals/#first-contentful-paint-fcp) of the transaction must be at least `2` seconds (with an upper bound of `10` seconds to eliminate outliers).
- The span's duration must be at least `33%` of the FCP.
```

## Span Evidence

> The evidence for a Large Render Blocking Asset problem identifies four aspects:

Reviewer: ```suggestion
You can get additional information about your Large Render Blocking Asset problem by looking at four aspects in the "Span Evidence" section:
```

- Transaction name
- Asset URL
- Effect on FCP
- Effect on total transaction time

> View it by going to the Issues page in Sentry, selecting your project, clicking on the Large Render Blocking Asset error you want to examine, then scrolling down to the "Span Evidence" section in the "Details" tab.
> 

Reviewer: ```suggestion
View it by going to the **Issues** page in Sentry, selecting your project, clicking on the Large Render Blocking Asset error you want to examine, then scrolling down to the "Span Evidence" section in the "Details" tab.

```

![Large Render Blocking Asset issue span evidence](span-evidence.png)

## Fingerprinting

> Sentry computes the issue fingerprint based on the asset URL. We detect many different patterns of cache-busting hashes in URL paths, filenames, and query strings so that the same asset is grouped into a single issue even as its contents changes.
> 

Reviewer: ```suggestion
Sentry computes the issue fingerprint based on the asset URL. We detect many different patterns of cache-busting hashes in URL paths, filenames, and query strings so that the same asset is grouped into a single issue even as its contents change.

```

## Recommendations

> Use tools like [Chrome's DevTools](https://developer.chrome.com/docs/devtools/) or [Firefox Developer Tools](https://firefox-dev.tools) to identify and extract non-critical CSS and JS in your pages. You can learn more about fixing these assets from the Chrome Developers post [Eliminate render-blocking resources](https://developer.chrome.com/en/docs/lighthouse/performance/render-blocking-resources/), and web.dev articles on [extracting critical CSS](https://web.dev/extract-critical-css/) and [deferring non-critical CSS](https://web.dev/defer-non-critical-css/)

Reviewer: ```suggestion
We recommend that you use tools like [Chrome's DevTools](https://developer.chrome.com/docs/devtools/) or [Firefox Developer Tools](https://firefox-dev.tools) to identify and extract non-critical CSS and JS in your pages. You can learn more about fixing these assets by reading the Chrome Developers post: [Eliminate render-blocking resources](https://developer.chrome.com/en/docs/lighthouse/performance/render-blocking-resources/), and web.dev articles on [extracting critical CSS](https://web.dev/extract-critical-css/) and [deferring non-critical CSS](https://web.dev/defer-non-critical-css/).
```

---
title: "Large Render Blocking Asset"
sidebar_order: 30
redirect_from:
  - /product/issues/performance-issues/large-render-blocking-asset/
description: "Learn more about large render blocking assets and how to diagnose and fix them."
---

<Include name="early-adopter-note.mdx" />

> A _Large Render Blocking Asset_ is a performance problem that happens when a large asset blocks the rendering of your web page, causing a delay in the page's paint. For example, when a page includes a large stylesheet, rendering of the page may stop until that stylesheet is fully downloaded and processed by the browser.

Reviewer 2: I think this would be even better with some more details about the impact and cause!

```suggestion
A _Large Render Blocking Asset_ is a performance problem that happens when a large asset causes a delay in displaying the page content. For example, if a page includes a stylesheet, the browser will pause rendering of the page until it downloads and processes the entire stylesheet. If the stylesheet is large, the user may see a blank or unstyled screen for a long time. This problem occurs when applications do not split up asset files, load non-essential assets synchronously, or load assets too early.
```

## Detection Criteria

The detector for this performance issue looks for specific asset (resource) spans with the following criteria:

> - The span has to be one of either: `resource.link` or `resource.script` spans.

Reviewer 2: ```suggestion
- The span operation has to be one of either: `resource.link` or `resource.script`
```

- The span's `resource.render_blocking_status` data property must have a value of `blocking` (added by the `@sentry/tracing` SDK as of version `7.38.0`) or be missing.

> - It has to be over `1mb` in size.

Reviewer 2: ```suggestion
- The span's `Encoded Body Size` data property has to be over `1MB` in size
```

- The [FCP (First Contentful Paint)](/product/performance/web-vitals/#first-contentful-paint-fcp) of the transaction must be at least `2` seconds (with an upper bound of `10` seconds to eliminate outliers).

> - The resource span must take up at least `33%` of the FCP.

Reviewer 2: ```suggestion
- The span's duration must be at least `33%` of the FCP
```

## Span Evidence

The evidence for a Large Render Blocking Asset problem identifies four aspects:

- Transaction name
- Asset URL
- Effect on FCP

> - Effect on total transaction time

Reviewer 2: Would be really good to include a screenshot here!

Author: 100%, had intended to and forgot - will certainly add one!

Author: @gggritso added!

Reviewer 2: Beautiful!

View it by going to the Issues page in Sentry, selecting your project, clicking on the Large Render Blocking Asset error you want to examine, then scrolling down to the "Span Evidence" section in the "Details" tab.

## Fingerprinting

Sentry computes the issue fingerprint based on the asset URL. We detect many different patterns of cache-busting hashes in URL paths, filenames, and query strings so that the same asset is grouped into a single issue even as its contents changes.

## Recommendations

> Use tools like Chrome's DevTools to identify non-critical CSS and JS in your render-blocking assets, and move the non-critical content into a separate non-blocking asset.
> 
> You can learn more about this approach from the following resource:
> 
> * [Lighthouse: Eliminate render-blocking resources](https://developer.chrome.com/en/docs/lighthouse/performance/render-blocking-resources/)

Reviewer 2: Call me anti-Google, but a mention of Firefox goes a long way 😉 Also adding some more resource links.

```suggestion
Use tools like [Chrome's DevTools](https://developer.chrome.com/docs/devtools/) or [Firefox Developer Tools](https://firefox-dev.tools) to identify and extract non-critical CSS and JS in your pages. You can learn more about fixing these assets from the Chrome Developers post [Eliminate render-blocking resources](https://developer.chrome.com/en/docs/lighthouse/performance/render-blocking-resources/), and web.dev articles on [extracting critical CSS](https://web.dev/extract-critical-css/) and [deferring non-critical CSS](https://web.dev/defer-non-critical-css/)
```
